File: esw-ocs/esw-ocs-dsl/src/main/scala/esw/dsl/script/epic/internal/Machine.scala
package esw.ocs.impl.dsl.epic.internal

import akka.Done
import esw.ocs.impl.dsl.epic.internal.event.Utils
import esw.ocs.impl.dsl.epic.{ProgramContext, Refreshable}

import scala.concurrent.Future
import scala.concurrent.duration.FiniteDuration
import scala.language.implicitConversions

abstract class Machine[State](name: String, init: State)(implicit programContext: ProgramContext) extends Refreshable {
  import programContext._
  type Logic = State => Unit

  def logic: Logic

  private var currentState: State  = init
  private var previousState: State = _

  implicit lazy val Refreshable: Refreshable = this

  protected def become(state: State): Unit = {
    currentState = state
  }

  def refresh(source: String): Future[Done] = {
    Future {
      println(
        f"machine = $name%-8s    previousState = $previousState%-8s     currentState = $currentState%-8s    action = $source%-8s     $debugString%8s"
      )
      logic(currentState)
      Done
    }
  }

  def when(condition: => Boolean = true)(body: => Unit): Unit = {
    previousState = currentState
    if (condition) {
      body
      refresh("when")
    }
  }

  def when(delay: FiniteDuration)(body: => Unit): Unit = {
    Utils.timeout(delay, strandEc.executorService).onComplete { _ =>
      when() {
        body
      }
    }
  }

  def entry(body: => Unit): Unit = {
    if (currentState != previousState) {
      body
    }
  }

  def debugString: String = ""

  implicit def varToT[T](reactive: Var[T]): T = reactive.get
}

object Machine {
  type Logic[T] = T => Unit
}

File: script-dsl/src/main/kotlin/esw/ocs/dsl/epics/Machine.kt
package esw.ocs.dsl.epics

import csw.params.core.generics.Parameter
import csw.params.core.generics.ParameterSetType
import csw.params.events.EventKey
import csw.params.events.SystemEvent
import esw.ocs.dsl.highlevel.EventServiceDsl
import esw.ocs.dsl.params.KeyKt
import esw.ocs.dsl.utils.CswExtensions
import esw.ocs.dsl.utils.nullable
import java.util.concurrent.Executors
import kotlin.coroutines.CoroutineContext
import kotlin.properties.ObservableProperty
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty
import kotlin.time.Duration
import kotlinx.coroutines.*

abstract class Machine(private val name: String, init: String) : CoroutineScope,
    Refreshable, EventServiceDsl, CswExtensions {

    private val ec = Executors.newSingleThreadScheduledExecutor()
    private val job = Job()
    private val dispatcher = ec.asCoroutineDispatcher()

    override val coroutineContext: CoroutineContext
        get() = job + dispatcher

    private var currentState: String = init
    private var previousState: String? = null

    abstract suspend fun logic(state: String)

    protected fun become(state: String) {
        currentState = state
    }

    override suspend fun refresh(source: String) {
        println(
            "machine = $name    previousState = $previousState     currentState = $currentState    action = $source     ${debugString()}"
        )
        logic(currentState)
    }

    suspend fun `when`(condition: Boolean = true, body: suspend () -> Unit) {
        previousState = currentState
        if (condition) {
            body()
            refresh("when")
        }
    }

    suspend fun `when`(duration: Duration, body: suspend () -> Unit) {
        delay(duration.toLongMilliseconds())
        `when`(body = body)
    }

    fun entry(body: () -> Unit) {
        if (currentState != previousState) {
            body()
        }
    }

    fun <T> Var(initial: T, eventKey: String, key: KeyKt<T>) = Var(initial, eventKey, this, this, key)

    inline fun <T> reactiveEvent(
        initial: T,
        eventKey: String,
        keyKt: KeyKt<T>,
        crossinline onChange: (oldValue: T, newValue: T) -> Unit
    ): ReadWriteProperty<Any?, T> =
        object : ObservableProperty<T>(initial) {
            private val _eventKey = EventKey.apply(eventKey)

            init {
                runBlocking {
                    publishEvent(event(keyKt.set(initial)))
                }
            }

            // todo: should allow creating any type of event
            private fun event(param: Parameter<T>): SystemEvent =
                SystemEvent(_eventKey.source(), _eventKey.eventName()).add(param)

            override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) =
                onChange(oldValue, newValue)

            override fun getValue(thisRef: Any?, property: KProperty<*>): T =
                runBlocking(coroutineContext) {
                    val paramType: ParameterSetType<*>? = getEvent(eventKey).first().paramType()
                    paramType?.jGet(keyKt.key)?.nullable()?.jGet(0)?.nullable() ?: initial
                }

            override fun setValue(thisRef: Any?, property: KProperty<*>, value: T) =
                runBlocking(coroutineContext) {
                    publishEvent(event(keyKt.set(value)))
                    super.setValue(thisRef, property, value)
                }
        }

    // todo: can we use generics here?
    operator fun Int?.compareTo(other: Int?): Int =
        if (this != null && other != null) this.compareTo(other)
        else -1

    open fun debugString(): String = ""
}
